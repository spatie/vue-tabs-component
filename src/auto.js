import Vue from 'vue';
import Tab from './components/Tab';
import Tabs from './components/Tabs';

Vue.component('tab', Tab);
Vue.component('tabs', Tabs);
