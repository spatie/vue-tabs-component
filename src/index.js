import Tab from './components/Tab';
import Tabs from './components/Tabs';
import Storages from './storages';

export default {
    install(Vue) {
        Vue.component('tab', Tab);
        Vue.component('tabs', Tabs);
    },
};

export { Tab, Tabs, Storages };
