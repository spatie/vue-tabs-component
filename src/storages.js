export function on(target, event, callback, capture = false) {
    target.addEventListener(event, callback, capture);
    return { destroy: () => {
        target.removeEventListener(event, callback, capture);
    } };
}

export class ExpiringStorage {
    constructor(options) {
        this.options = options;
    }

    createKey() {
        const prefix = typeof this.options.prefix === 'string'
            ? this.options.prefix
            : 'vue-tabs-component.cache.';
        return `${prefix}${this.options.key}`;
    }

    get() {
        const key = this.createKey();
        const cached = JSON.parse(
            localStorage.getItem(key)
        );

        if (! cached) {
            return null;
        }

        const expires = new Date(cached.expires);

        if (expires < new Date()) {
            localStorage.removeItem(key);
            return null;
        }

        return cached.value;
    }

    set(value) {
        const key = this.createKey();
        const currentTime = new Date().getTime();
        const expires = new Date(currentTime + this.options.cacheLifeTimeInMinutes * 60000);

        localStorage.setItem(key, JSON.stringify({ value, expires }));
    }
}

export class HashStorage {
    constructor(options) {
        this.options = options;
        this.unmount = HashStorage.mount(this);
    }

    get() {
        return window.location.hash ? window.location.hash : null;
    }

    set(value) {
        if (this.options.useUrlFragment && value !== this.get()) {
            window.location.hash = value;
        }
    }

    destroy() {
        this.unmount();
    }

    static dispatchChange() {
        for (const instance of this.instances) {
            instance.options.autoSelectTab();
        }
    }

    static mount(instance) {
        if (typeof this.instances === 'undefined') {
            this.instances = [];
        }

        if (typeof this.listener === 'undefined') {
            this.listener = on(window, 'hashchange', this.dispatchChange.bind(HashStorage));
        }

        this.instances.push(instance);
        return () => {
            const index = this.instances.indexOf(instance);

            if (index > -1) {
                this.instances.splice(index, 1);
            }

            if (this.instances.length === 0) {
                this.listener.destroy();
                delete this.listener;
            }
        };
    }
}

export class ChainStorage {
    constructor(options) {
        this.options = options;
    }

    get() {
        for (const backend of this.options.backends) {
            const value = backend.get();

            if (value !== null) {
                return value;
            }
        }

        return null;
    }

    set(value) {
        const continueOnSuccess = this.options.continueOnSuccess || true;
        let result = true;

        for (const backend of this.options.backends) {
            const wasSuccessful = backend.set(value) !== false;
            result &= wasSuccessful;

            if (wasSuccessful && !continueOnSuccess) {
                return result;
            }
        }

        return result;
    }

    destroy() {
        for (const backend of this.options.backends) {
            backend.destroy();
        }
    }
}

export default {
    ExpiringStorage,
    HashStorage,
    ChainStorage,
};
