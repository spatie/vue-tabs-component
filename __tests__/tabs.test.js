import { Tab, Tabs } from '../src';
import Vue from 'vue/dist/vue.js';
import expiringStorage from '../src/expiringStorage';
import LocalStorageMock from './helpers/LocalStorageMock';

const localStorage = new LocalStorageMock();

window.localStorage = localStorage;

describe('vue-tabs-component', () => {
    Vue.component('tabs', Tabs);
    Vue.component('tab', Tab);

    beforeEach(() => {
        document.body.innerHTML = `
            <div id="app">
                <tabs>
                    <tab name="First tab">
                        First tab content
                    </tab>
                    <tab name="Second tab">
                        Second tab content
                    </tab>
                    <tab name="Third tab">
                        Third tab content
                    </tab>
                </tabs>
            </div>
        `;

        localStorage.clear();

        const dateClass = Date;

        // eslint-disable-next-line no-global-assign
        Date = function (dateString) {
            return new dateClass(dateString || '2017-01-01T00:00:00.000Z');
        };

        window.location.hash = '';
    });

    it('can mount tabs', async () => {
        await createVm();

        expect(document.body.innerHTML).toMatchSnapshot();
    });

    it('displays the first tab by default', async () => {
        const tabs = await createVm();

        expect(tabs.activeTabHash).toEqual('#first-tab');
    });

    it('uses a custom fragment', async () => {
        document.body.innerHTML = `
            <div id="app">
                <tabs cache-lifetime="10">
                    <tab id="my-fragment" name="First tab" >
                        First tab content
                    </tab>
                </tabs>
            </div>
        `;

        const tabs = await createVm();

        expect(tabs.activeTabHash).toEqual('#my-fragment');
    });

    it('uses the fragment of the url to determine which tab to open', async () => {
        window.location.hash = '#second-tab';

        const tabs = await createVm();

        expect(document.body.innerHTML).toMatchSnapshot();
    });

    it('ignores the fragment if it does not match the hash of a tab', async () => {
        window.location.hash = '#unknown-tab';

        const tabs = await createVm();

        expect(tabs.activeTabHash).toEqual('#first-tab');
    });

    it('writes the hash of the last opened tab in local storage', async () => {
        window.location.hash = '#third-tab';

        const tabs = await createVm();

        expect(tabs.activeTabHash).toEqual('#third-tab');

        expect(localStorage.getAll()).toMatchSnapshot();
    });

    it('opens up the tabname found in storage', async () => {
        expiringStorage.set('vue-tabs-component.cache.blank', '#third-tab', 5);

        const tabs = await createVm();

        expect(tabs.activeTabHash).toEqual('#third-tab');
    });

    it('will not use the tab in storage if it has expired', async () => {
        expiringStorage.set('vue-tabs-component.cache.blank', '#third-tab', 5);

        progressTime(6);

        const tabs = await createVm();

        expect(tabs.activeTabHash).toEqual('#first-tab');
    });

    it('the life time of the cache can be set', async () => {
        document.body.innerHTML = `
            <div id="app">
                <tabs cache-lifetime="10">
                    <tab name="First tab">
                        First tab content
                    </tab>
                </tabs>
            </div>
        `;

        await createVm();

        expect(localStorage.getAll()).toMatchSnapshot();
    });

    it('can accept a prefix and a suffix for the name', async () => {
        document.body.innerHTML = `
            <div id="app">
                <tabs cache-lifetime="10">
                    <tab name="First tab" prefix="prefix" suffix="suffix">
                        First tab content
                    </tab>
                </tabs>
            </div>
        `;

        await createVm();

        expect(document.body.innerHTML).toMatchSnapshot();
    });

    it('nested removes the tab hash', async () => {
        document.body.innerHTML = `
            <div id="app">
                <tabs nested>
                    <tab name="First tab" >
                        First tab content
                    </tab>
                </tabs>
            </div>
        `;

        await createVm();

        expect(document.body.innerHTML).toMatchSnapshot();
    });

    it('moves to the correct tab with wrapping false', async () => {
        const tabs = await createVm();

        // test nextTab/previousTab with wrapping false
        expect(tabs.activeTabHash).toEqual('#first-tab');
        tabs.nextTab();
        expect(tabs.activeTabHash).toEqual('#second-tab');
        tabs.nextTab();
        expect(tabs.activeTabHash).toEqual('#third-tab');
        tabs.nextTab();
        expect(tabs.activeTabHash).toEqual('#third-tab');
        tabs.previousTab();
        expect(tabs.activeTabHash).toEqual('#second-tab');
        tabs.previousTab();
        expect(tabs.activeTabHash).toEqual('#first-tab');
        tabs.previousTab();
        expect(tabs.activeTabHash).toEqual('#first-tab');
    });

    it('moves to the correct tab with wrapping true', async () => {
        const tabs = await createVm();

        // test nextTab/previousTab with wrapping true
        expect(tabs.activeTabHash).toEqual('#first-tab');
        tabs.nextTab(true);
        expect(tabs.activeTabHash).toEqual('#second-tab');
        tabs.nextTab(true);
        expect(tabs.activeTabHash).toEqual('#third-tab');
        tabs.nextTab(true);
        expect(tabs.activeTabHash).toEqual('#first-tab');

        tabs.previousTab(true);
        expect(tabs.activeTabHash).toEqual('#third-tab');
        tabs.previousTab(true);
        expect(tabs.activeTabHash).toEqual('#second-tab');
        tabs.previousTab(true);
        expect(tabs.activeTabHash).toEqual('#first-tab');
    });

    it('gets and sets tabs by index number', async () => {
        const tabs = await createVm();

        expect(tabs.getActiveTabIndex()).toEqual(0);
        tabs.selectTabByIndex(1);
        expect(tabs.activeTabHash).toEqual('#second-tab');
        expect(tabs.getActiveTabIndex()).toEqual(1);
        tabs.nextTab();
        expect(tabs.getActiveTabIndex()).toEqual(2);
    });
});

async function createVm() {
    const vm = new Vue({
        el: '#app',
    });

    await Vue.nextTick();

    return vm.$children[0];
}

function progressTime(minutes) {
    const currentTime = (new Date()).getTime();

    const newTime = new Date(currentTime + (minutes * 60000));

    const originalDateClass = Date;

    // eslint-disable-next-line no-global-assign
    Date = function (dateString) {
        return new originalDateClass(dateString || newTime.toISOString());
    };
}
