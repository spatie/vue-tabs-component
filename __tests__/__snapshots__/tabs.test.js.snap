// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`vue-tabs-component can accept a prefix and a suffix for the name 1`] = `
"
            <div id=\\"app\\"><div class=\\"tabs-component\\"><ul role=\\"tablist\\" class=\\"tabs-component-tabs\\"><li role=\\"presentation\\" class=\\"tabs-component-tab is-active\\"><a aria-controls=\\"#first-tab\\" aria-selected=\\"true\\" href=\\"#first-tab\\" role=\\"tab\\" class=\\"tabs-component-tab-a\\">prefixFirst tabsuffix</a></li></ul> <div class=\\"tabs-component-panels\\"><section id=\\"#first-tab\\" role=\\"tabpanel\\" class=\\"tabs-component-panel\\" style=\\"\\">
                        First tab content
                    </section></div></div></div>
        "
`;

exports[`vue-tabs-component can mount tabs 1`] = `
"
            <div id=\\"app\\"><div class=\\"tabs-component\\"><ul role=\\"tablist\\" class=\\"tabs-component-tabs\\"><li role=\\"presentation\\" class=\\"tabs-component-tab is-active\\"><a aria-controls=\\"#first-tab\\" aria-selected=\\"true\\" href=\\"#first-tab\\" role=\\"tab\\" class=\\"tabs-component-tab-a\\">First tab</a></li><li role=\\"presentation\\" class=\\"tabs-component-tab\\"><a aria-controls=\\"#second-tab\\" href=\\"#second-tab\\" role=\\"tab\\" class=\\"tabs-component-tab-a\\">Second tab</a></li><li role=\\"presentation\\" class=\\"tabs-component-tab\\"><a aria-controls=\\"#third-tab\\" href=\\"#third-tab\\" role=\\"tab\\" class=\\"tabs-component-tab-a\\">Third tab</a></li></ul> <div class=\\"tabs-component-panels\\"><section id=\\"#first-tab\\" role=\\"tabpanel\\" class=\\"tabs-component-panel\\" style=\\"\\">
                        First tab content
                    </section> <section aria-hidden=\\"true\\" id=\\"#second-tab\\" role=\\"tabpanel\\" class=\\"tabs-component-panel\\" style=\\"display: none;\\">
                        Second tab content
                    </section> <section aria-hidden=\\"true\\" id=\\"#third-tab\\" role=\\"tabpanel\\" class=\\"tabs-component-panel\\" style=\\"display: none;\\">
                        Third tab content
                    </section></div></div></div>
        "
`;

exports[`vue-tabs-component nested removes the tab hash 1`] = `
"
            <div id=\\"app\\"><div class=\\"tabs-component\\"><ul role=\\"tablist\\" class=\\"tabs-component-tabs\\"><li role=\\"presentation\\" class=\\"tabs-component-tab is-active\\"><a aria-controls=\\"#first-tab\\" aria-selected=\\"true\\" role=\\"tab\\" class=\\"tabs-component-tab-a\\">First tab</a></li></ul> <div class=\\"tabs-component-panels\\"><section id=\\"#first-tab\\" role=\\"tabpanel\\" class=\\"tabs-component-panel\\" style=\\"\\">
                        First tab content
                    </section></div></div></div>
        "
`;

exports[`vue-tabs-component the life time of the cache can be set 1`] = `
Object {
  "vue-tabs-component.cache.blank": "{\\"value\\":\\"#first-tab\\",\\"expires\\":\\"2017-01-01T00:10:00.000Z\\"}",
}
`;

exports[`vue-tabs-component uses the fragment of the url to determine which tab to open 1`] = `
"
            <div id=\\"app\\"><div class=\\"tabs-component\\"><ul role=\\"tablist\\" class=\\"tabs-component-tabs\\"><li role=\\"presentation\\" class=\\"tabs-component-tab\\"><a aria-controls=\\"#first-tab\\" href=\\"#first-tab\\" role=\\"tab\\" class=\\"tabs-component-tab-a\\">First tab</a></li><li role=\\"presentation\\" class=\\"tabs-component-tab is-active\\"><a aria-controls=\\"#second-tab\\" aria-selected=\\"true\\" href=\\"#second-tab\\" role=\\"tab\\" class=\\"tabs-component-tab-a\\">Second tab</a></li><li role=\\"presentation\\" class=\\"tabs-component-tab\\"><a aria-controls=\\"#third-tab\\" href=\\"#third-tab\\" role=\\"tab\\" class=\\"tabs-component-tab-a\\">Third tab</a></li></ul> <div class=\\"tabs-component-panels\\"><section aria-hidden=\\"true\\" id=\\"#first-tab\\" role=\\"tabpanel\\" class=\\"tabs-component-panel\\" style=\\"display: none;\\">
                        First tab content
                    </section> <section id=\\"#second-tab\\" role=\\"tabpanel\\" class=\\"tabs-component-panel\\" style=\\"\\">
                        Second tab content
                    </section> <section aria-hidden=\\"true\\" id=\\"#third-tab\\" role=\\"tabpanel\\" class=\\"tabs-component-panel\\" style=\\"display: none;\\">
                        Third tab content
                    </section></div></div></div>
        "
`;

exports[`vue-tabs-component writes the hash of the last opened tab in local storage 1`] = `
Object {
  "vue-tabs-component.cache.blank": "{\\"value\\":\\"#third-tab\\",\\"expires\\":\\"2017-01-01T00:05:00.000Z\\"}",
}
`;
